File: solutions/javascript/protein-translation/1/protein-translation.js
import { throwStatement } from "@babel/types";

//
// This is only a SKELETON file for the 'Protein Translation' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const translate = (rna) => {
    var protein_sequence = [];

    if (rna !== undefined) {
        var l = rna.length;

        while (l > 0) {
            let codon = rna.slice(0,3);
            l = l-3;
            rna = rna.slice(-(l));

            switch(codon) {
                case 'AUG':
                    protein_sequence.push('Methionine');
                    break;
                case 'UUU':
                    protein_sequence.push('Phenylalanine');
                    break;
                case 'UUC':
                    protein_sequence.push('Phenylalanine');
                    break;
                case 'UUA':
                    protein_sequence.push('Leucine');
                    break;
                case 'UUG':
                    protein_sequence.push('Leucine');
                    break;
                case 'UCU':
                    protein_sequence.push('Serine');
                    break;
                case 'UCC':
                    protein_sequence.push('Serine');
                    break;
                case 'UCA':
                    protein_sequence.push('Serine');
                    break;
                case 'UCG':
                    protein_sequence.push('Serine');
                    break;
                case 'UAU':
                    protein_sequence.push('Tyrosine');
                    break;
                case 'UAC':
                    protein_sequence.push('Tyrosine');
                    break;
                case 'UGU':
                    protein_sequence.push('Cysteine');
                    break;
                case 'UGC':
                    protein_sequence.push('Cysteine');
                    break;
                case 'UGG':
                    protein_sequence.push('Tryptophan');
                    break;
                case 'UAA': // STOP
                    return protein_sequence;
                case 'UAG': // STOP
                    return protein_sequence;
                case 'UGA': // STOP
                    return protein_sequence;
                default:
                    throw('Invalid codon');
            }
        }
    }

    return protein_sequence;
};

File: solutions/go/protein-translation/2/protein_translation.go
package protein

import (
	"errors"
)

// ErrInvalidBase blah
var ErrInvalidBase = errors.New("InvalidBase")

// ErrStop blah
var ErrStop = errors.New("STOP")

// FromCodon converts a codon to it's human readable protein name
func FromCodon(codon string) (string, error) {
	protein := ""

	switch codon {
	case "AUG":
		protein = "Methionine"
		break
	case "UUU":
		protein = "Phenylalanine"
		break
	case "UUC":
		protein = "Phenylalanine"
		break
	case "UUA":
		protein = "Leucine"
		break
	case "UUG":
		protein = "Leucine"
		break
	case "UCU":
		protein = "Serine"
		break
	case "UCC":
		protein = "Serine"
		break
	case "UCA":
		protein = "Serine"
		break
	case "UCG":
		protein = "Serine"
		break
	case "UAU":
		protein = "Tyrosine"
		break
	case "UAC":
		protein = "Tyrosine"
		break
	case "UGU":
		protein = "Cysteine"
		break
	case "UGC":
		protein = "Cysteine"
		break
	case "UGG":
		protein = "Tryptophan"
		break
	case "UAA":
		return "", ErrStop
	case "UAG":
		return "", ErrStop
	case "UGA":
		return "", ErrStop
	default:
		return "", ErrInvalidBase
	}

	return protein, nil
}

// FromRNA converts a RNA to it's human readable protein name by translating it's codons
func FromRNA(rna string) ([]string, error) {
	var proteins []string

	if len(rna)%3 != 0 {
		return proteins, ErrInvalidBase
	}

	for i := 0; i < len(rna); i = i + 3 {
		protein, err := FromCodon(rna[i : i+3])

		if err == nil {
			proteins = append(proteins, protein)
		} else if err == ErrStop {
			return proteins, nil
		} else {
			return proteins, err
		}
	}
	return proteins, nil
}
